### PR TITLE
[YUNIKORN-2190] Using empty array to replace null response from /ws/v1/partition/{partitionName}/usage/users

### PR DESCRIPTION
**Pull request by brandboat**
### What is this PR for?
The endpoint `/ws/v1/partition/{partitionName}/usage/users` and `/ws/v1/partition/{partitionName}/usage/groups`
will return nil when user / group usages are nonexistent. The return type is expected to be array, so it seems to me the empty array is more suitable to be returned.

### What type of PR is it?
* [x] - Improvement

### Todos
N/A

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2190

### How should this be tested?
covered by unit tests

### Screenshots (if appropriate)
N/A

### Questions:
N/A

### CODE REVIEW

File: pkg/webservice/handlers.go [hunk 1]
```diff
@@ -823,7 +823,7 @@ func getUsersResourceUsage(w http.ResponseWriter, _ *http.Request) {
 	writeHeaders(w)
 	userManager := ugm.GetUserManager()
 	usersResources := userManager.GetUsersResources()
-	var result []*dao.UserResourceUsageDAOInfo
+	result := []*dao.UserResourceUsageDAOInfo{}
```
Comment (chia7712) [inline pkg/webservice/handlers.go:826]: how about using `make` with initial capacity? 

```
result := make([]*dao.UserResourceUsageDAOInfo, len(usersResources))
```

File: pkg/webservice/handlers.go [hunk 2]
```diff
@@ -859,7 +859,7 @@ func getGroupsResourceUsage(w http.ResponseWriter, r *http.Request) {
 	writeHeaders(w)
 	userManager := ugm.GetUserManager()
 	groupsResources := userManager.GetGroupsResources()
-	var result []*dao.GroupResourceUsageDAOInfo
+	result := []*dao.GroupResourceUsageDAOInfo{}
```
Comment (chia7712) [inline pkg/webservice/handlers.go:862]: ditto

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for this patch. a couple of comments below.

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (pbacsko) [reply pkg/webservice/handlers.go:826]: +1 for that

Comment (pbacsko) [reply pkg/webservice/handlers.go:826]: Actually I think it has to be `result := make([]*dao.UserResourceUsageDAOInfo, 0, len(usersResources))` so that appending starts at position 0.

Comment (wilfred-s) [reply pkg/webservice/handlers.go:826]: The even faster solution is to use this:
```
	result := make([]*dao.UserResourceUsageDAOInfo, len(usersResources))
	for i, tracker := range usersResources {
		result[i] = tracker.GetUserResourceUsageDAOInfo()
	}
```

Comment (brandboat) [reply pkg/webservice/handlers.go]: Already addressed comments, thank you for everyone's valuable suggestions !

File: pkg/webservice/handlers_test.go [hunk 3]
```diff
@@ -1500,7 +1500,7 @@ func TestUsersAndGroupsResourceUsage(t *testing.T) {
 	assert.NilError(t, err, "Get Groups Resource Usage Handler request failed")
 
 	var groupsResourceUsageDao []*dao.GroupResourceUsageDAOInfo
-	var expGroupsResourceUsageDao []*dao.GroupResourceUsageDAOInfo
+	expGroupsResourceUsageDao := []*dao.GroupResourceUsageDAOInfo{}
```
Comment (chia7712) [inline pkg/webservice/handlers_test.go:1503]: pardon me, is this change related to PR? or it is a potential issue?

**Review by chia7712 — APPROVED** [review]
overall +1, one small question left.

File: pkg/webservice/handlers_test.go (hunk 3 above)
Comment (brandboat) [reply pkg/webservice/handlers_test.go:1503]: Yes it's related. Since the response from the code you mentioned is now empty array instead of nil as we changed the behavior in this pr.

IMO, the test should be adjusted as follows: set up a scenario where the group resource usage is not empty, ensuring that there is content in groupsResourceUsageDao. Otherwise, it appears to be duplicating the functionality of TestEmptyGroupsResourceUsageHandler.

Comment (chia7712) [reply pkg/webservice/handlers_test.go:1503]: > IMO, the test should be adjusted as follows: set up a scenario where the group resource usage is not empty, ensuring that there is content in groupsResourceUsageDao. Otherwise, it appears to be duplicating the functionality of TestEmptyGroupsResourceUsageHandler.

that makes sense to me. Please make sure the test is NOT duplicate to `TestEmptyGroupsResourceUsageHandler`

Comment (manirajv06) [reply pkg/webservice/handlers_test.go:1503]: @brandboat Empty response is one of the basic test cases while hitting those API's. So, can we add the test related changes in `TestUsersAndGroupsResourceUsage` itself instead of two new & separate test methods?

Comment (brandboat) [reply pkg/webservice/handlers_test.go]: Already addressed comments in the latest commit, many thanks !

File: pkg/webservice/handlers_test.go [hunk 4]
```diff
@@ -1405,7 +1422,7 @@ func TestFullStateDumpPath(t *testing.T) {
 }
 
 func TestSpecificUserAndGroupResourceUsage(t *testing.T) {
-	prepareUserAndGroupContext(t)
+	prepareUserAndGroupContext(t, configDefault)
```
Comment (chia7712) [inline pkg/webservice/handlers_test.go:1425]: (this is not related to this PR)
it seems the `configDefault` does not include any group/user, so all asserts in this test case are related to "nonexistent" case, right?

**Review by pbacsko — APPROVED** [review]
+1 LGTM

File: pkg/webservice/handlers_test.go (hunk 4 above)
Comment (brandboat) [reply pkg/webservice/handlers_test.go:1425]: > it seems the configDefault does not include any group/user, so all asserts in this test case are related to "nonexistent" case, right?

You're correct, except for the test case under comment `// Test existed user query`. I believe the test case mentioned in that comment is missing an assertion to verify the existence of the user resource `testuser`.

Comment (chia7712) [reply pkg/webservice/handlers_test.go:1425]: That is unrelated to this PR, so I file ticket (https://issues.apache.org/jira/browse/YUNIKORN-2272) to trace it.

**Review by chia7712 — APPROVED** [review]
LGTM

File: pkg/webservice/handlers_test.go (hunk 4 above)
Comment (brandboat) [reply pkg/webservice/handlers_test.go:1425]: Thanks, already take over the ticket, will come up with another pr ASAP.
